oss-internship-2020/guetzli/guetzli_transaction.cc

```diff
@@ -0,0 +1,120 @@
+#include "guetzli_transaction.h"
+
+#include <iostream>
+#include <memory>
+
+namespace guetzli {
+namespace sandbox {
+
+absl::Status GuetzliTransaction::Init() {
+  // Close remote fd if transaction is repeated
+  if (in_fd_.GetRemoteFd() != -1) {
+    SAPI_RETURN_IF_ERROR(in_fd_.CloseRemoteFd(sandbox()->GetRpcChannel()));
+  }
+  if (out_fd_.GetRemoteFd() != -1) {
+    SAPI_RETURN_IF_ERROR(out_fd_.CloseRemoteFd(sandbox()->GetRpcChannel()));
+  }
+
+  // Reposition back to the beginning of file
+  if (lseek(in_fd_.GetValue(), 0, SEEK_CUR) != 0) {
+    if (lseek(in_fd_.GetValue(), 0, SEEK_SET) != 0) {
+      return absl::FailedPreconditionError(
+        "Error returnig cursor to the beginning"
+      );
+    }
+  }
+
+  // Choosing between jpg and png modes
+  sapi::StatusOr<ImageType> image_type = GetImageTypeFromFd(in_fd_.GetValue());
+
+  if (!image_type.ok()) {
+    return image_type.status();
+  }
+
+  image_type_ = image_type.value();
+
+  SAPI_RETURN_IF_ERROR(sandbox()->TransferToSandboxee(&in_fd_));
+  SAPI_RETURN_IF_ERROR(sandbox()->TransferToSandboxee(&out_fd_));
+
+  if (in_fd_.GetRemoteFd() < 0) {
+    return absl::FailedPreconditionError(
+        "Error receiving remote FD: remote input fd is set to -1");
+  }
+  if (out_fd_.GetRemoteFd() < 0) {
+    return absl::FailedPreconditionError(
+        "Error receiving remote FD: remote output fd is set to -1");
+  }
+
+  in_fd_.OwnLocalFd(false); // FDCloser will close local fd
+  out_fd_.OwnLocalFd(false); // FDCloser will close local fd
+
+  return absl::OkStatus();
+}
+
+absl::Status GuetzliTransaction::Main() {
+  GuetzliApi api(sandbox());
+  sapi::v::LenVal output(0);
+
+  sapi::v::Struct<ProcessingParams> processing_params;
+  *processing_params.mutable_data() = {in_fd_.GetRemoteFd(), 
+                                      params_.verbose, 
+                                      params_.quality, 
+                                      params_.memlimit_mb
+  };
+
+  auto result_status = image_type_ == ImageType::JPEG ? 
+    api.ProcessJpeg(processing_params.PtrBefore(), output.PtrBoth()) : 
+    api.ProcessRgb(processing_params.PtrBefore(), output.PtrBoth());
+  
+  if (!result_status.value_or(false)) {
```

Comment:
use `SAPI_ASSIGN_OR_RETURN`
distinguish between sandbox error and library error